File: python-service/services/chunker.py
"""
Text chunking service.
Splits page-level text into overlapping chunks that respect
sentence and paragraph boundaries.
"""

import re


def chunk_by_page(
    pages: list[dict],
    chunk_size: int = 1000,
    overlap: int = 200,
) -> list[dict]:
    """
    Split extracted page texts into overlapping chunks.

    Args:
        pages: List of dicts with 'page_number' and 'content' keys
               (as returned by pdf_extractor.extract_text_by_page).
        chunk_size: Target chunk size in characters.
        overlap: Number of overlapping characters between consecutive chunks.

    Returns:
        List of dicts with keys: page_number, chunk_index, content.
    """
    all_chunks = []

    for page in pages:
        page_number = page["page_number"]
        # Strip NUL bytes that some PDFs contain — Postgres UTF-8 columns
        # reject them and the whole chunk insert fails.
        text = (page["content"] or "").replace("\x00", "")

        if not text.strip():
            continue

        page_chunks = _split_text(text, chunk_size, overlap)

        for idx, chunk_text in enumerate(page_chunks):
            # Double-check after the split — _split_text returns substrings
            # of the input but belt + braces against any stray NULs.
            cleaned = chunk_text.replace("\x00", "")
            if not cleaned.strip():
                continue
            all_chunks.append({
                "page_number": page_number,
                "chunk_index": idx,
                "content": cleaned,
            })

    return all_chunks


def _split_text(text: str, chunk_size: int, overlap: int) -> list[str]:
    """
    Split text into chunks of approximately chunk_size characters,
    with overlap characters of overlap between consecutive chunks.
    Splits respect sentence and paragraph boundaries where possible.
    """
    if len(text) <= chunk_size:
        return [text]

    sentences = _split_into_sentences(text)
    chunks = []
    current_chunk: list[str] = []
    current_length = 0

    for sentence in sentences:
        sentence_len = len(sentence)

        # If a single sentence exceeds chunk_size, force-split it
        if sentence_len > chunk_size:
            # Flush current chunk first
            if current_chunk:
                chunks.append(" ".join(current_chunk))
                current_chunk = []
                current_length = 0

            # Hard split the long sentence
            for i in range(0, sentence_len, chunk_size - overlap):
                fragment = sentence[i : i + chunk_size]
                chunks.append(fragment)
            continue

        # If adding this sentence would exceed chunk_size, flush
        if current_length + sentence_len + 1 > chunk_size and current_chunk:
            chunk_text = " ".join(current_chunk)
            chunks.append(chunk_text)

            # Build overlap from the tail of the current chunk
            overlap_text = _build_overlap(current_chunk, overlap)
            current_chunk = [overlap_text] if overlap_text else []
            current_length = len(overlap_text) if overlap_text else 0

        current_chunk.append(sentence)
        current_length += sentence_len + 1  # +1 for space

    # Don't forget the last chunk
    if current_chunk:
        chunks.append(" ".join(current_chunk))

    return chunks


def _split_into_sentences(text: str) -> list[str]:
    """
    Split text into sentences, respecting paragraph boundaries.
    Uses a regex-based approach that handles common abbreviations.
    """
    # First split on paragraph boundaries
    paragraphs = re.split(r"\n\s*\n", text)
    sentences = []

    for para in paragraphs:
        para = para.strip()
        if not para:
            continue

        # Split on sentence-ending punctuation followed by space or end
        # Handles: . ! ? and their combinations with quotes/parens
        parts = re.split(
            r'(?<=[.!?])\s+(?=[A-Z\u00C0-\u024F\u0400-\u04FF"\'\(])',
            para,
        )

        for part in parts:
            part = part.strip()
            if part:
                sentences.append(part)

    return sentences


def _build_overlap(chunks: list[str], target_overlap: int) -> str:
    """
    Build an overlap string from the tail of a list of sentences,
    aiming for approximately target_overlap characters.
    """
    if not chunks or target_overlap <= 0:
        return ""

    result_parts = []
    total = 0

    for sentence in reversed(chunks):
        if total + len(sentence) > target_overlap and result_parts:
            break
        result_parts.insert(0, sentence)
        total += len(sentence) + 1

    return " ".join(result_parts)

File: src/lib/chunker.ts
/**
 * Text chunking — node port of python-service/services/chunker.py.
 *
 * Splits per-page text into overlapping chunks that respect sentence
 * and paragraph boundaries. Used by the library extraction pipeline
 * once it pulls page text via pdfjs (src/lib/pdf-extract.ts).
 *
 * Kept ~byte-identical to the Python implementation so we can roll
 * back to the Python service if anything regresses — the chunk-level
 * embedding output is downstream of this and won't notice as long as
 * the cleanup + split boundaries match.
 */

export interface ChunkerInputPage {
  pageNumber: number;
  pageLabel?: string | null;
  content: string;
}

export interface ChunkerOutput {
  pageNumber: number;
  pageLabel: string | null;
  chunkIndex: number;
  content: string;
}

export interface ChunkOptions {
  chunkSize?: number;
  overlap?: number;
}

/**
 * Split each page's text into overlapping chunks.
 *
 * Default sizing mirrors the Python defaults so embedding token
 * budgets stay aligned.
 */
export function chunkByPage(
  pages: ChunkerInputPage[],
  options: ChunkOptions = {},
): ChunkerOutput[] {
  const chunkSize = options.chunkSize ?? 1000;
  const overlap = options.overlap ?? 200;
  const out: ChunkerOutput[] = [];

  for (const page of pages) {
    // Postgres UTF-8 columns reject NUL bytes; strip both at page-text
    // level and after the split (belt + braces — _split_text returns
    // substrings of the input but anything leaking through here kills
    // a whole batch insert).
    const text = (page.content ?? "").replace(/\u0000/g, "");
    if (!text.trim()) continue;

    const pieces = splitText(text, chunkSize, overlap);
    for (let i = 0; i < pieces.length; i++) {
      const cleaned = pieces[i].replace(/\u0000/g, "");
      if (!cleaned.trim()) continue;
      out.push({
        pageNumber: page.pageNumber,
        pageLabel: page.pageLabel ?? null,
        chunkIndex: i,
        content: cleaned,
      });
    }
  }

  return out;
}

function splitText(text: string, chunkSize: number, overlap: number): string[] {
  if (text.length <= chunkSize) return [text];

  const sentences = splitIntoSentences(text);
  const chunks: string[] = [];
  let current: string[] = [];
  let currentLen = 0;

  for (const sentence of sentences) {
    const sLen = sentence.length;

    // If a single sentence exceeds chunkSize, flush what we have and
    // hard-split the long sentence at chunkSize windows.
    if (sLen > chunkSize) {
      if (current.length > 0) {
        chunks.push(current.join(" "));
        current = [];
        currentLen = 0;
      }
      const step = Math.max(1, chunkSize - overlap);
      for (let i = 0; i < sLen; i += step) {
        chunks.push(sentence.slice(i, i + chunkSize));
      }
      continue;
    }

    // Adding this sentence would overflow — emit the current chunk
    // first, then start a new one seeded with overlap from the tail.
    if (currentLen + sLen + 1 > chunkSize && current.length > 0) {
      chunks.push(current.join(" "));
      const overlapText = buildOverlap(current, overlap);
      current = overlapText ? [overlapText] : [];
      currentLen = overlapText.length;
    }

    current.push(sentence);
    currentLen += sLen + 1; // +1 for the joining space
  }

  if (current.length > 0) chunks.push(current.join(" "));
  return chunks;
}

function splitIntoSentences(text: string): string[] {
  // First split on paragraph boundaries (blank lines).
  const paragraphs = text.split(/\n\s*\n/);
  const out: string[] = [];
  for (const rawPara of paragraphs) {
    const para = rawPara.trim();
    if (!para) continue;
    // Sentence boundary: punctuation + whitespace + uppercase-or-
    // open-quote. Matches the Python regex (covers Latin uppercase,
    // Latin-1 Supplement, Latin Extended, Cyrillic, plus opening
    // quotes/parens — handles common abbreviations OK).
    const parts = para.split(
      /(?<=[.!?])\s+(?=[A-ZÀ-ɏЀ-ӿ"'(])/u,
    );
    for (const part of parts) {
      const p = part.trim();
      if (p) out.push(p);
    }
  }
  return out;
}

function buildOverlap(chunks: string[], target: number): string {
  if (chunks.length === 0 || target <= 0) return "";
  const result: string[] = [];
  let total = 0;
  for (let i = chunks.length - 1; i >= 0; i--) {
    const s = chunks[i];
    if (total + s.length > target && result.length > 0) break;
    result.unshift(s);
    total += s.length + 1;
  }
  return result.join(" ");
}
